src/util/math.h

```diff
@@ -38,40 +45,38 @@ inline bool even(T value) {
 #pragma intrinsic(fabs)
 #endif
 
-inline int roundUpToPowerOf2(int v) {
-    int power = 1;
-    while (power < v && power > 0) {
+constexpr unsigned int roundUpToPowerOf2(int v) {
+    DEBUG_ASSERT(v >= 0);
+#ifdef __cpp_lib_bitops
+    const auto uv = static_cast<unsigned int>(v);
+    return std::bit_ceil(uv);
+#else
+    unsigned int power = 1;
```

Comment:
This needs to be a signed that the power > 0 condition works.
But does it actually work?

Reply:
No this does not work with signed ints. The reason is that overflow/underflow in signed ints is UB. The same is not true of unsigned ints, so when those overflow, they will become 0 and the loop will terminate.
https://stackoverflow.com/questions/16188263/is-signed-integer-overflow-still-undefined-behavior-in-c
Here is a godbolt that shows that the previous version can end up in an infinite loop while the newer version will actually work.https://compiler-explorer.com/z/ae5P9edGE
The previous version also had the bug that it would use the next power of two if what was passed in already was a power of two.

Reply:
The reason why signed into overflow is undefined In C++ is that this depends on the target hardware. With exotic CPUs ther is no guarantee that sign bit is the MSB. Luckily for all Mixxx targets this is the case, and we rely on it in some other places.

Reply:
The point is when we're relying on the overflow, we're relying on UB and so the compiler is free to do whatever it wants (which is does and thus creates an implementation that could end up in an infinite loop).

Reply:
If you now pass the maximum signed int of 0x7fffffff, it returns the unsigned int of 0x80000000
Such value will break the using code, because it is interpreted as -1. int32_t is used as ring_buffer_size_t.

This considerations are only valid on target hardware with the MSB sign, so we can rely here on the integer overflow behavior. If you not trust it, we can add a static assertion to verify it.

So I recommend to reintroduce the original check ans use a signed type in both cases.

Reply:
It is also surprising that a rounding function also changes the signess  of the value it should round.

Reply:
> It is also surprising that a rounding function also changes the signess of the value it should round.

The original function never even worked for negative values. Also its really only used in a **single** place (fifo.h which is why the same commit contains changes there). The function should've always taken an unsigned int in the first place. but for some reason, mixxx is obsessed with using signed types when they are inappropriate (`typedef SINT ptrdiff_t` being the most prominent example). `DEBUG_ASSERT(v >= 0);` being the most obvious indicator of this.